providers/certificate.rb

```diff
@@ -0,0 +1,160 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# See this for info on certutil
+# https://technet.microsoft.com/en-gb/library/cc732443.aspx
+
+include Chef::Mixin::ShellOut
+include Chef::Mixin::PowershellOut
+include Windows::Helper
+
+# Support whyrun
+def whyrun_supported?
+  true
+end
+
+action :create do
+	# we create from a file...
+	file = win_friendly_path(@new_resource.source)
+	
+	hash = getHashFromFile(file)
+	if isCertInStore?(hash)
+		Chef::Log.debug("#{@new_resource.source} already exists - nothing to do")
+	else
+		converge_by("Adding #{ @new_resource.source }") do
+			addCertToStore!(file)
+			updateAcl!(hash)
+		end
+	end
+
+end
+
+# acl_add is a modify-if-exists operation : not idempotent
+action :acl_add do
+	hash = nil
+
+	if ::File.exists?(@new_resource.source)
+		# source is a file so get hash from that
+		file = win_friendly_path(@new_resource.source)
+		hash = getHashFromFile(file)
+	else
+		# make sure we have no spaces in the hash string
+		hash = @new_resource.source.gsub(/\s/, '')
+	end
+	
+	if isCertInStore?(hash)
+		converge_by("Adding to ACL of #{hash}") do
+			updateAcl!(hash)
+		end
+	else
+		Chef::Log.debug("#{@new_resource.source} does not exist - nothing to do")
+	end
+end
+
+action :delete do
+	# on delete the source should id the cert (subject, hash, serial number etc.)
+	if isCertInStore?(@new_resource.source)
+		converge_by("Deleting #{ @current_resource.source }") do
+			user = " -user" if @new_resource.user_store
+			cmd = shell_out!("#{@command}#{user} -delstore #{@new_resource.store_name} \"#{@current_resource.source}\"")
+		end
+	else
+		Chef::Log.debug("#{@new_resource.source} does not exist - nothing to do")
+	end
+end
+
+def load_current_resource
+	@current_resource = Chef::Resource::WindowsCertificate.new(@new_resource.name)
+	@current_resource.source(@new_resource.source)
+	@current_resource.pfx_password(@new_resource.pfx_password)
+	@current_resource.private_key_acl(@new_resource.private_key_acl)
+	@current_resource.store_name(@new_resource.store_name)
+	@current_resource.user_store(@new_resource.user_store)
+
+	@command = locate_sysnative_cmd("certutil.exe")
+end
+
+private
+def getHashFromFile(file)
+	if file.downcase.end_with?('pfx') && (@new_resource.pfx_password.nil? || @new_resource.pfx_password.empty?)
+		raise "No password given for PFX file"
+	end
+	
+	password = " -p #{@new_resource.pfx_password}" if file.downcase.end_with?('pfx')
+	cmd = shell_out("#{@command}#{password} \"#{file}\"")
+	Chef::Log.debug "certutil reports: #{cmd.stdout}"
+	
+	# extract values from returned text
+	if cmd.exitstatus == 0
```

Comment:
If you're already using powershell features below, you can make some of this cleaner by using (New-Object System.Security.Cryptography.X509Certificates.X509Certificate "file").GetCertHashString() in a powershell_script.  Actually, now that I think of it, a lot of this can be simplified by using powershell_script and putting the checks in guard clauses.